File: benchmarking/tests/passport_processing_01/solution/solution.py
import re

# define the set of mandatory fields
mandatory_fields = {
    'byr': [(re.compile('^([0-9]{4})$'), 1920, 2002)],
    'iyr': [(re.compile('^([0-9]{4})$'), 2010, 2020)],
    'eyr': [(re.compile('^([0-9]{4})$'), 2020, 2030)],
    'hgt': [(re.compile('^([0-9]+)cm$'), 150, 193), (re.compile('^([0-9]+)in$'), 59, 78)],
    'hcl': [(re.compile('^#[0-9a-f]{6}$'), None, None)],
    'ecl': [(re.compile('^amb|blu|brn|gry|grn|hzl|oth$'), None, None)],
    'pid': [(re.compile('^[0-9]{9}$'), None, None)],
}

def is_valid_passport(passport):

    """
    >>> is_valid_passport('ecl:gry pid:860033327 eyr:2020 hcl:#fffffd byr:1937 iyr:2017 cid:147 hgt:183cm')
    True
    >>> is_valid_passport('iyr:2013 ecl:amb cid:350 eyr:2023 pid:028048884 hcl:#cfa07d byr:1929')
    False
    >>> is_valid_passport('hcl:#ae17e1 iyr:2013 eyr:2024 ecl:brn pid:760753108 byr:1931 hgt:179cm')
    True
    >>> is_valid_passport('hcl:#cfa07d eyr:2025 pid:166559648 iyr:2011 ecl:brn hgt:59in')
    False
    """

    # extract the set of fields from the passport
    fields = {field.split(':')[0]:field.split(':')[1] for field in passport.split()}

    # check if all mandatory fields are present
    return set(mandatory_fields) <= set(fields)

def passports(filename):

    """
    >>> list(passports('passports.txt'))
    ['ecl:gry pid:860033327 eyr:2020 hcl:#fffffd byr:1937 iyr:2017 cid:147 hgt:183cm', 'iyr:2013 ecl:amb cid:350 eyr:2023 pid:028048884 hcl:#cfa07d byr:1929', 'hcl:#ae17e1 iyr:2013 eyr:2024 ecl:brn pid:760753108 byr:1931 hgt:179cm', 'hcl:#cfa07d eyr:2025 pid:166559648 iyr:2011 ecl:brn hgt:59in']
    """

    with open(filename) as lines:
        passport = []
        for line in lines:
            line = line.rstrip('\n')
            if line:
                passport.append(line)
            elif passport:
                yield ' '.join(passport)
                passport = []

        if passport:
            yield ' '.join(passport)

def count_valid_passports(filename):

    """
    >>> count_valid_passports('passports.txt')
    2
    >>> count_valid_passports('adventofcode.input.txt')
    247
    """

    # count the number of valid passports in the given file
    return sum(is_valid_passport(passport) for passport in passports(filename))

if __name__ == '__main__':
    import doctest
    doctest.testmod()

File: benchmarking/tests/passport_processing_01/solution/solution.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define INC_RIGHT 3

bool is_valid_passport(const char *pass)
{
    return strstr(pass, "byr:")
        && strstr(pass, "iyr:")
        && strstr(pass, "eyr:")
        && strstr(pass, "hgt:")
        && strstr(pass, "hcl:")
        && strstr(pass, "ecl:")
        && strstr(pass, "pid:");
}

int count_valid_passports(const char *filename)
{
    FILE *file = fopen(filename, "r");

    // Include space for newline and string terminator
    char buffer[128] = { 0 };

    bool has_byr = false;
    bool has_iyr = false;
    bool has_eyr = false;
    bool has_hgt = false;
    bool has_hcl = false;
    bool has_ecl = false;
    bool has_pid = false;
    int correct = 0;
    while (fgets(buffer, 128, file)) {
        if (buffer[0] == '\n') {
            if (has_byr && has_iyr && has_eyr && has_hgt && has_hcl && has_ecl && has_pid) {
                correct++;
            }

            has_byr = false;
            has_iyr = false;
            has_eyr = false;
            has_hgt = false;
            has_hcl = false;
            has_ecl = false;
            has_pid = false;
        }

        has_byr = has_byr || strstr(buffer, "byr:");
        has_iyr = has_iyr || strstr(buffer, "iyr:");
        has_eyr = has_eyr || strstr(buffer, "eyr:");
        has_hgt = has_hgt || strstr(buffer, "hgt:");
        has_hcl = has_hcl || strstr(buffer, "hcl:");
        has_ecl = has_ecl || strstr(buffer, "ecl:");
        has_pid = has_pid || strstr(buffer, "pid:");
    }

    if (has_byr && has_iyr && has_eyr && has_hgt && has_hcl && has_ecl && has_pid) {
        correct++;
    }

    fclose(file);

    return correct;
}
